davidia/server/plotserver.py

```diff
@@ -226,9 +414,74 @@ async def handle_client(server: PlotServer, plot_id: str, socket: WebSocket):
                     omit = client  # omit originating client
 
                 # currently used to test websocket communication in test_api
-                server.prepare_data(received_message, omit_client=omit)
+                await server.prepare_data(received_message, omit_client=omit)
                 await server.send_next_message()
 
     except WebSocketDisconnect:
         logger.error("Websocket disconnected:", exc_info=True)
         server.remove_client(plot_id, client)
+
+
+class PlotState:
+    """Class for representing current data messages."""
+
+    def __init__(
+        self,
+        new_data_message=None,
+        new_selections_message=None,
+        current_data=None,
+        current_selections=None,
+    ):
+        self.new_data_message: bytes | None = new_data_message
+        self.new_selections_message: bytes | None = new_selections_message
+        self.current_data: DataMessage | None = current_data
+        self.current_selections: SelectionsMessage | None = current_selections
+        self.lock = Lock()
+
+    def clear(self):
+        """Clear all current and new data and selections"""
+        self.new_data_message = None
+        self.new_selections_message = None
+        self.current_data = None
+        self.current_selections = None
+
+
+def add_indices(msg: MultiLineDataMessage) -> MultiLineDataMessage:
+    """Adds default indices to a multi-line data message if default indices flag is True
+
+    Parameters
+    ----------
+    msg : MultiLineDataMessage
+        A multi-line data message to which to add indices.
+
+    Returns the new multi-line data message
+    """
+    if msg.ml_data[0].default_indices:
+        for m in msg.ml_data:
+            m.x = np.arange(m.y.size, dtype=m.y.dtype)
```

Comment:
Don't use `y`'s dtype as it could be very limited. Use https://numpy.org/doc/stable/reference/generated/numpy.min_scalar_type.html for all invocations of `arange` where you are generating indexes.